### DISCUSSION TITLE
Looking for Great Python Learning Resources

### DISCUSSION THREAD

**Discussion by MalikIssiakhem** [community/community#170006] [Programming Help]
Hi everyone! 👋
I'm currently deepening my understanding of Python and looking to explore more advanced or structured resources beyond the basics. I’ve gone through some tutorials and beginner-friendly documentation, but I’d love to hear what others recommend for really leveling up.
Whether it's:
Free or paid online courses
YouTube channels or playlists
Books (digital or physical)
Practice platforms (e.g., for coding challenges or projects)
Blogs or newsletters
Open source projects that are great for learning by contributing
Anything else you’ve found useful
Feel free to drop your suggestions below! Bonus points if the resource focuses on things like best practices, real-world projects, or intermediate-to-advanced topics (like async, testing, architecture, etc.).
Thanks in advance for sharing! 🙏

**Comment by ghost** [comment]
I think this is the best place to learn python 3
https://www.codecademy.com/learn/learn-python-3
<img width="1150" height="381" alt="Screenshot 2025-08-16 151814" src="https://github.com/user-attachments/assets/166f5509-d895-4011-9c49-0e445faf79ad" />

**Reply by MalikIssiakhem** [reply]
Sorry, I didn’t take the time to reply earlier. Thank you for your time and for the resources you shared!

**Comment by Rosellines** [comment]
Great question! 🚀 Once you’re past the basics, the best way to level up is to dive into best practices + real-world projects. Here are some resources I recommend:

Books

Fluent Python by Luciano Ramalho → idiomatic & advanced Python.

Effective Python by Brett Slatkin → 90+ practical tips.

Courses / Online

Real Python → excellent paid tutorials for intermediate/advanced learners.

CS50’s Python track (free) → project-based learning.

Talk Python Training → async, FastAPI, testing, and more.

YouTube

[ArjanCodes](https://www.youtube.com/@ArjanCodes) → best practices & architecture.

[Socratica](https://www.youtube.com/@Socratica) → concise, clear explanations.

[Tech With Tim](https://www.youtube.com/@TechWithTim) → project tutorials.

Practice Platforms

Exercism Python Track → free, mentored challenges.

LeetCode → sharpen algorithms & problem-solving.

Advent of Code → fun puzzles in December.

Blogs / Newsletters

PyCoders Weekly → weekly roundup.

Real Python Blog → deep dives & tutorials.

Open Source to Learn From

[Django](https://github.com/django/django) → classic large-scale project.

[FastAPI](https://github.com/tiangolo/fastapi) → modern async Python web framework.

If you want to focus on best practices, I’d suggest pairing Effective Python with [ArjanCodes](https://www.youtube.com/@ArjanCodes), then practice by contributing to something like [FastAPI](https://github.com/tiangolo/fastapi).

**Comment by tuves** [comment]
Welcome to the GitHub Community, @Malouek, we're happy you're here! You are more likely to get a useful response if you are posting your question(s) in the applicable category and are explicit about what your project entails--giving a few more details might help someone give you a nudge in the right direction. I've gone ahead and moved it for you. Good luck!

**Comment by wislertt** [comment]
[leetcode-py](https://github.com/wisarootl/leetcode-py/) check this, if you want to practice leetcode using python.

**Reply by MalikIssiakhem** [reply]
Sorry, I didn’t take the time to reply earlier. Thank you for your time and for the resources you shared!

**Comment by tstenborg** [comment]
The [Python Institute](https://pythoninstitute.org/) has some good learning resources.

**Reply by MalikIssiakhem** [reply]
Sorry, I didn’t take the time to reply earlier. Thank you for your time and for the resources you shared!

**Comment by Sahil-dev7** [comment]
https://roadmap.sh/python

The Best technique was that where, you Teach yourself by own.

**Reply by MalikIssiakhem** [reply]
Sorry, I didn’t take the time to reply earlier. Thank you for your time and for the resources you shared!

**Comment by Venkatsai04** [comment]
First step—_escape_ tutorial hell. Don’t just follow along; start building with a purpose. Ask yourself why you’re learning Python (web, data, AI, automation), then pick resources that match that goal.

**Some solid next steps:**

**Books**: Fluent Python (Luciano Ramalho), Effective Python (Brett Slatkin)

**YouTube**: Corey Schafer (great for intermediate concepts), Tech With Tim

**Courses**: “Python Asyncio” on RealPython, “Testing with pytest” (TestDriven.io)

**Practice**: LeetCode, Codewars, Project Euler (for problem-solving)

**Open Source:** Check issues labeled good first issue on Django, FastAPI, or Pandas

**Purpose** + **practice** > endless tutorials. That’s how you level up.

**Reply by MalikIssiakhem** [reply]
Sorry, I didn’t take the time to reply earlier. Thank you for your time and for the resources you shared!

**Comment by gpl-gowthamchand** [comment]
Hey @Malouek,

That’s a really good move - looking for structured, advanced-level Python resources will take you further than just tutorials. Here are some suggestions, and how you can make the most out of them:

---

### Resources I Recommend

* *Fluent Python* by Luciano Ramalho - great for thinking in more “Python-ic” ways.
* *Effective Python* by Brett Slatkin - lots of practical tips that help you write cleaner, more robust code.
* RealPython (website) - their intermediate/advanced tutorials on testing, async, web frameworks are solid.
* LeetCode, Exercism, Advent of Code - for building problem solving and algorithmic skills.
* Open source frameworks like FastAPI or Django - try contributing to small issues (“good first issue”) to see how production-level code is organized.

---

### How to Make Your Learning More Effective

What helped me when I was leveling up:

* Pick a project that interests you (web app, automation, data work, etc.). Use the resources *while you build*.
* Read code from real, active projects. Look at how folks structure folders, how they write tests, how they do error handling.
* Write tests early - even for small features. Testing forces you to think about edge cases and behavior.
* Review your own code after a few weeks/months; see what you could do better now that your understanding is deeper.
* Join communities, watch code reviews, ask for feedback.

**Reply by MalikIssiakhem** [reply]
Sorry, I didn’t take the time to reply earlier. Thank you for your time and for the resources you shared!

**Comment by nikhiilraj** [comment]
Here are some Python resources that really helped me level up beyond the basics:
Courses:
Intermediate Python (free) and Talk Python’s courses (paid but very practical).
YouTube:
Corey Schafer for real-world explanations, and ArjanCodes for clean code + architecture.
Books:
Fluent Python for deeper concepts and Effective Python for best practices.
Practice:
LeetCode or HackerRank for challenges, and Exercism if you want mentor-style feedback.
Blogs/Newsletters:
RealPython tutorials and the Python Weekly newsletter.
Open Source:
Projects like Django, FastAPI, or Pandas have “good first issue” tags that are great for learning.
These helped me move from just writing Python to actually understanding how to build cleaner, more real-world code. Hope they help you too!

**Comment by GanLin-06** [comment]
Hi there! 🚀 Great goal—leveling up Python means focusing on real-world practices and advanced concepts. Here’s a condensed, high-impact list of resources:
📚 Must-Read Books
Fluent Python (2nd Ed.): The gold standard for writing idiomatic, efficient Python—dives into internals (decorators, context managers, data models).
Python Cookbook (3rd Ed.): Practical recipes for complex tasks (async, metaprogramming) with clear explanations.
Effective Python (3rd Ed.): 90+ actionable guidelines for best practices (type hints, error handling, performance).
🖥️ Online Courses (Free & Paid)
Paid: Udemy’s Python for Professionals (Mosh Hamedani) – skips basics, covers advanced topics (async, design patterns).
Free: Real Python’s Intermediate/Advanced Tutorials – deep, code-heavy articles on decorators, generators, type hints.
🎥 YouTube Channels
Corey Schafer: Advanced Python playlist (decorators, threading, async I/O) – clear, practical examples.
Tech With Tim: Python projects & advanced tutorials (API building, testing, concurrency).
🧩 Practice Platforms
Exercism: Focuses on code quality/reviews (great for refining best practices).
LeetCode/HackerRank: Python-specific challenges (strengthens problem-solving with advanced syntax).
Real Python Projects: Step-by-step real-world builds (e.g., APIs, data pipelines).
🌐 Open Source Learning
Search GitHub for projects tagged good first issue (e.g., Python good first issue) – beginner-friendly contributions to learn collaboration.
PyTorch/TensorFlow Docs: If into ML, their tutorials blend Python best practices with real use cases.
Quick Tip: Prioritize projects (e.g., build a Flask API, data scraper with async) – applying concepts beats passive learning! Happy coding! 🙏

**Comment by Xishen1823** [comment]
You can find excellent non - video Python resources through these channels: First, rely on authoritative official docs like Python.org Official Tutorials and Microsoft’s Python Learning Path for accurate and updated basic and advanced knowledge. Then, turn to well - structured written tutorials such as freeCodeCamp’s The Python Handbook and Liao Xuefeng’s Chinese Python tutorial, which combine theory with practical examples. For systematic learning, choose text - focused courses like Coursera’s Python for Everybody and Chinese University MOOC’s Python courses by prestigious domestic universities, which come with downloadable slides and exercise materials.
 
In terms of practical improvement, refer to professional blogs like RealPython that offer in - depth guides on Web development and data analysis. Dive into open - source projects on GitHub, such as Awesome Python and The Algorithms - Python, to learn from high - quality code. You can also enhance your skills through coding challenge platforms like LeetCode and HackerRank by solving Python algorithm problems. Besides, classic books are great helpers—beginners can start with Python Crash Course, advanced learners can opt for Fluent Python, and those into data analysis can choose Python for Data Analysis. Finally, communities like Stack Overflow and Reddit’s r/learnpython let you get answers to problems and stay updated on industry trends.

**Comment by evee00011** [comment]
When I first started learning Python, I focussed on learning the basic programming concepts first. There are things like what a programming language is and what is its purpose and really looking at the big picture. 
After that, I followed a YouTube tutorial by Programming by Mosh. I'd suggest to not focus too much on completing the YouTube tutorial since hands on experience is 100% more valuable. So after gaining a brief introduction, move on to doing small projects to learn. 
I personally used Python Institute and went the certification route. I did the PCAP certification which is an amazing beginner level cert. Their courses on Python Institute were free the last I remember and they were really effective with lots of hands on practice.

**Comment by Maqbool61** [comment]
Great timing to dive deeper into Python! Here's a structured path based on what actually works:

## Free resources (high signal)

**Reading**
- [Real Python](https://realpython.com) — deep, code-heavy articles on decorators, generators, async, type hints
- [Python Docs — "What's New"](https://docs.python.org/3/whatsnew/) — read one per version, teaches you how the language evolves

**Video**
- Corey Schafer on YouTube — best advanced Python playlist (decorators, threading, context managers)
- ArjanCodes — excellent for design patterns and clean architecture in Python

## Books worth buying
- *Fluent Python* (Ramalho) — the single best book for going from intermediate to advanced
- *Architecture Patterns with Python* (Percival & Gregory) — free online, teaches DDD/hexagonal architecture

## Practice
- [Exercism](https://exercism.org/tracks/python) — community code reviews on your solutions (huge for learning style)
- Contribute to a real open-source project — search GitHub for `label:good-first-issue language:python`

## One tip
Pick **one** resource and finish it rather than collecting tutorials. *Fluent Python* cover-to-cover will teach you more than 20 half-finished courses.
